### DISCUSSION TITLE
Using a union for the discriminator with discriminatedUnions fails

### DISCUSSION THREAD

**Discussion by affanshahid** [colinhacks/zod#2776] [Q&A] [answered]
Using a union for the discriminator value with discriminatedUnions yields this runtime error: 

```
Error: A discriminator value for key `status` could not be extracted from all schema options
```

Sample code:
```ts
const myUnion = z.discriminatedUnion("status", [
  z.object({ status: z.union([z.literal("success"), z.literal("ok")]) , data: z.string() }),
  z.object({ status: z.literal("failed"), error: z.instanceof(Error) }),
]);

myUnion.parse({ status: "success", data: "yippie ki yay" });
```

The valid type for this would be something like:

```ts
type myUnion = {
  status: "success" | "ok";
  data: string;
} | {
  status: "failed";
  error: Error;
}
```

The workaround is to duplicate the the entry for each value in the inner union.

**Comment by conorbrandon** [comment]
I _just_ ran into this same issue (the case of using a `ZodUnion` discriminator comprised of only `ZodLiteral`s). 

I went here to file an issue and found this was the latest one. Perfect timing! 🤯 

1. Would it be feasible to allow `ZodUnion`s with only `ZodLiteral`s as a valid discriminator option? I totally understand if not, because another workaround (@affanshahid) is to use a `ZodEnum`, which has the exact same effect as `ZodUnion<ZodLiteral[]>`.
2. Maybe instead, the discriminator schema in [`ZodDiscriminatedUnionOption`](https://github.com/colinhacks/zod/blob/1ecd6241ef97b33ce229b49f1346ffeee5d0ba74/src/types.ts#L2895) could be restricted from `ZodTypeAny` to one of the _actual_ allowed schemas as seen in [`getDiscriminator`](https://github.com/colinhacks/zod/blob/1ecd6241ef97b33ce229b49f1346ffeee5d0ba74/src/types.ts#L2870)?
3. Perhaps, the [error message](https://github.com/colinhacks/zod/blob/1ecd6241ef97b33ce229b49f1346ffeee5d0ba74/src/types.ts#L3001):
    1. Could be improved to indicate exactly which schema was unable to be extracted, instead of just saying "all"?
    2. Include as part of the message which schemas are even allowed as discriminators?
4. The discriminated union [docs](https://zod.dev/?id=discriminated-unionshttps://zod.dev/?id=discriminated-unions) could make a note about common pitfalls? (i.e. which schemas are even allowed as discriminators)

I really enjoy using Zod, and many thanks to all who've contributed to make it great, but I feel there could be an improvement made here somewhere... (At least in the documentation? 🥲)

**Comment by affanshahid** [comment]
Thanks for the enum suggestion @conorbrandon. I can't use the enum approach it seems since it only accepts a list of strings and in my actual use case the discriminator is a number.

**Comment by conorbrandon** [comment]
Ah, gotcha!

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const myUnion = z.discriminatedUnion( 'status', [
    z.object( {
        status: z.literal( 'success' ),
        data: z.string(),
    } ),
    z.object( {
        status: z.literal( 'ok' ),
        data: z.string(),
    } ),
    z.object( {
        status: z.literal( 'failed' ),
        error: z.instanceof( Error ),
    } ),
] )

myUnion.parse( { status: 'success', data: 'yippie ki yay' } )
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by Anoesj** [comment]
Hmm, what a shame it only supports `z.enum`, but not `z.union`. Is there a way to get around this without specifying all literals in separate objects if the literals are numbers?
